src/js/formapp.js

```diff
@@ -106,15 +105,53 @@ function getSubmission(formData, formSubmission, reducers, evalContext) {
   });
 }
 
-async function renderForm({ definition, formData, formSubmission, reducers, contextScripts }) {
+let prevSubmission;
+let isChanging = false;
+
+const hasChangedFunction = 'return function hasChanged(key) { return !_.isEqual(_.get(formSubmission, key), _.get(prevSubmission, key)); }';
+
+const onChange = debounce(function(form, changeReducers) {
+  isChanging = true;
+
+  const data = reduce(changeReducers, (memo, reducer) => {
+    const context = form.evalContext({ formSubmission: memo, prevSubmission });
+    context.hasChanged = FormioUtils.evaluate(hasChangedFunction, context);
+    return FormioUtils.evaluate(reducer, context) || memo;
+  }, structuredClone(form.submission.data));
+
+  form.setSubmission({ data });
+
+  prevSubmission = structuredClone(form.submission.data);
+
+  isChanging = false;
+}, 300);
```

Comment:
What's the purpose of the `300` milliseconds wait before calling the `onChange()` function?

Not questioning the choice, just curious what it does.

Reply:
I think I got the number from the debounce that form.io is doing on their change event, but essentially what this does is allow for a bunch of changes while only calling onChange once every 300ms.. So for instance if we removed this we'd run all of the onChanges on every keystroke inside of an input on a form.  It might be better to be a bit quicker actually, but for now it should mirror current response times on data changes.  As opposed to throttle, it's triggered at the end of the time so that you always get the latest change when the event is triggered.

Reply:
Interesting... okay, thanks!